### ISSUE TITLE
RMSE logging when training data contains molecules of varying sizes

### ISSUE THREAD

**Issue by ojaace** [atomistic-machine-learning/schnetpack#24]
I'm unsure whether or not I'm getting representative RMSE logging when training on a data set that contains molecules of varying sizes (from 55 atoms to 190 atoms).

If I'm not mistaken, the following piece of code is run (per batch?) when logging the RMSE of a target:

https://github.com/atomistic-machine-learning/schnetpack/blob/master/src/schnetpack/metrics.py#L68

```python
    def add_batch(self, batch, result):
        y = batch[self.target]
        if self.model_output is None:
            yp = result
        else:
            yp = result[self.model_output]

        diff = self._get_diff(y, yp)
        self.l2loss += torch.sum(diff.view(-1)).detach().cpu().data.numpy()
        self.n_entries += np.prod(y.shape)
```

https://github.com/atomistic-machine-learning/schnetpack/blob/master/src/schnetpack/metrics.py#L132

```python
    def aggregate(self):
        return np.sqrt(self.l2loss / self.n_entries)
```
As far as I understand, there's some zero padding happening in the background when the training batch contains molecules of varying sizes. Am I correct in assuming that this zero padding increases the number ``self.n_entries`` (when using ``np.prod(y.shape)``) such that my mean (for the smaller molecules) becomes a lot smaller than it should be?

**Comment by ktschuett** [comment]
If you are predicting energies, there is no zero-padding involved. However, if you are prediction per-atom properties, you are correct and we need to address this. Is that what you are doing?

I guess the best way to deal with this is to extend the metrics, so one can pass the padding mask.

**Comment by ojaace** [comment]
My loss function contains both the energy and the per-atom forces, and I am tracking the RMSE of both of these properties.

I stumbled across this when trying to find (and use) the mean of the per-atom forces in my loss function.

**Comment by ktschuett** [comment]
Ok, than you will have a problem with the force loss. Do you want to fix it? Otherwise, we will have have a look at it as soon as we find some time.

**Comment by ojaace** [comment]
We've been using ``batch['_atomic_numbers']`` to extract the number of atoms for use in the loss function, so I believe training is working properly. This is more of a logging issue.

I'll wait for a fix, I don't feel confident in my ability to solve the issue myself.

Thanks a lot for the very fast response. :)

**Comment by ojaace** [comment]
I've been looking at solving this issue. I've made an edit to set ``self.n_entries`` differently if the target is forces. There's probably a much better way to do this, but I believe it works properly.


The line

https://github.com/atomistic-machine-learning/schnetpack/blob/master/src/schnetpack/metrics.py#L112
  
has been replaced with:

```python
if self.target == "forces":
    N = torch.sum((batch['_atomic_numbers'] != 0).float(), 1, keepdim=True)
    N *= 3
    self.n_entries += np.sum(torch.sum(N).detach().cpu().data.numpy())
else:
    self.n_entries += np.prod(y.shape)
```

Let me know if there are any obvious errors!
